src/Access/SettingsConstraints.h

```diff
@@ -87,29 +82,60 @@ class SettingsConstraints
     friend bool operator !=(const SettingsConstraints & left, const SettingsConstraints & right) { return !(left == right); }
 
 private:
-    struct Constraint
+    enum ReactionOnViolation
+    {
+        THROW_ON_VIOLATION,
+        CLAMP_ON_VIOLATION,
+    };
+
+    struct Range
     {
-        std::shared_ptr<const String> setting_name;
-        bool read_only = false;
+        SettingConstraintType type = SettingConstraintType::NONE;
         Field min_value;
         Field max_value;
 
-        bool operator ==(const Constraint & other) const;
-        bool operator !=(const Constraint & other) const { return !(*this == other); }
+        String explain;
+        int code;
```

Comment:
It's just required to carry exception. Which in turn is required to completely incapsulate all logic into this class. Currently checks are duplicated in system table. These was required to get rid of that duplication

Reply:
Probably it was not clear enough. `SettingsConstraints::get` and `SettingsConstraints::checkImpl` are now both using `getRange` and operate with `Range` to share common logic: e.g. strange `allow_ddl` check to avoid duplicating this logic in system table code. But I wanted to keep all exception messages exactly the same, so I came up with this trick. Maybe you could propose some better way to achieve the goal?

Reply:
After all this is just helper class visible only inside implementation, so I guess it's okay